README.md
# [monitode](http://supergiovane.tk/#/monitode)

[![NPM version](https://badge.fury.io/js/monitode.svg)](http://badge.fury.io/js/monitode)
[![Build Status](https://travis-ci.org/hex7c0/monitode.svg?branch=master)](https://travis-ci.org/hex7c0/monitode)
[![Dependency Status](https://david-dm.org/hex7c0/monitode/status.svg)](https://david-dm.org/hex7c0/monitode)

[![monitode logo](https://raw.githubusercontent.com/hex7c0/monitode/master/public/img/logo.png)](https://hex7c0.github.io/monitode/)

Resource monitor for [nodejs](http://nodejs.org) using a full stack of [MEAN](http://en.wikipedia.org/wiki/MEAN).

## Installation

Install through NPM

```bash
npm install monitode
```
or
```bash
git clone git://github.com/hex7c0/monitode.git
```

## API

inside nodejs project
```js
var monitode = require('monitode')();
```

inside expressjs project
```js
var monitode = require('monitode');
var app = require('express')();

app.use(monitode({
    password : 'psw'
}));
```

inside expressjs project with [`logger-request`](https://github.com/hex7c0/logger-request)
```js
var monitode = require('monitode');
var app = require('express')();
var logger = require('logger-request');

var file = __dirname + '/monitode.log'
app.use(logger({
    filename : file,
}));
app.use(monitor({
    log : file,
}));
```

Monidote informations are stored inside _GLOBAL_ **Object**.
One istance for environment.
```js
GLOBAL.monitode;
```

### monitode(options)

#### options

 - `output` - **Boolean** If enabled, show output to shell console *(default "false")*
 - `os` - **Boolean** If enabled, show os (netstat, iostat) statistics to web console/database/email *(default "false")* **_NOT FOR WIN_**
 - `tickle` - **Boolean** If enabled, using [`tickle`](https://github.com/hex7c0/tickle) inside dynamic datas *(default "disabled")*
 - `app` - **Object** If enabled, using Express object for [`express-sitemap`](https://github.com/hex7c0/express-sitemap) inside static datas *(default "disabled")*
 
 
 
 - `http` - **Object** Setting up a web console using [`basic-authentication`](https://github.com/hex7c0/basic-authentication)
  - `enabled` - **Boolean** If disabled, don't run web console *(default "enabled")*
  - `port` - **Integer** Which port accept connection of web console *(default "30000")*
  - `user` - **String** User for web basic access authentication *(default "admin")*
  - `password` - **String** Password for web basic access authentication *(default "password")*
  - `agent` - **String** Browser User Agent for web authentication *(default "all accepted")*
  - `realm` - **String** Realm for web authentication *(default "Monitode")*
  - `file` - **String** Path of htpasswd file *(default "disabled")*
  - `hash` - **String** Type of [hash](http://nodejs.org/api/crypto.html#crypto_crypto_createhash_algorithm) inside your htpasswd file *(default "md5")*
  - `dir` - **String** Absolute path of web files directory *(default "/public")*
 
  related to https://github.com/hex7c0/basic-authentication
 
 - `https` - **Object** Setting up a web console over TLS/SSL using [`basic-authentication`](https://github.com/hex7c0/basic-authentication)
  - `key` - **String** Path to TLS/SSL key *(default "disabled")*
  - `cert` - **String** Path to TLS/SSL certificate *(default "disabled")*
  - `port` - **Integer** Which port accept connection of web console over TLS/SSL *(default "30003")*
  - `user` - **String** User for web basic access authentication over TLS/SSL *(default "admin")*
  - `password` - **String** Password for web basic access authentication over TLS/SSL *(default "password")*
  - `agent` - **String** Browser User Agent for web authentication over TLS/SSL *(default "all accepted")*
  - `realm` - **String** Realm for web authentication over TLS/SSL *(default "Monitode")*
  - `file` - **String** Path of htpasswd file *(default "disabled")*
  - `hash` - **String** Type of [hash](http://nodejs.org/api/crypto.html#crypto_crypto_createhash_algorithm) inside your htpasswd file *(default "md5")*
  - `dir` - **String** Absolute path of web files directory *(default "/public")*
 
  related to https://github.com/hex7c0/basic-authentication
 
 - `logger` - **Object** Setting up a permanent file which save your story
  - `log` - **String** Path to log file, using [`logger-request`](https://github.com/hex7c0/logger-request) *(default "disabled")* Parsing information stored inside log, and show it to web console or file/database
  - `file` - **String** Path to file *(default "disabled")* Save stats at regular intervals to file
  - `timeout` - **Float** Timeout (second) for file write *(default "5")*
 
  related to https://github.com/hex7c0/logger-request
 
 - `db` - **Object** Setting up a connectiont to database which save your story
  - `mongo` - **String** URI for MongoDb connection *(default "disabled")* Save stats at regular intervals to database
  - `couch` - **String** URI for CouchDb connection *(default "disabled")* Save stats at regular intervals to database
  - `database` - **String** Name of your database *(default "monitode")*
  - `timeout` - **Float** Timeout (second) for database query *(default "20")*
 
  related to https://github.com/mongodb/node-mongodb-native and https://github.com/dscape/nano
 
 - `mail` - **Object** Setting up SMTP
  - `provider` - **String** Check [`nodemailer`](https://github.com/andris9/nodemailer#well-known-services-for-smtp) for available email provider *(default "disabled")*
  - `user` - **String** User for email authentication *(default "admin")*
  - `password` - **String** Password for email authentication *(default "password")*
  - `to` - **Array** Write here your destination emails *(default "empty")*
  - `subject` - **String** Email subject *(default "password")*
  - `timeout` - **Float** Timeout (second) for email send *(default "60")*
 
 
 
 - `status` - **Object** Setting up a check of status of any number of websites and save status to file
  - `enabled` - **Boolean** If disabled, don't run check status *(default "disabled")*
  - `sites` - **Array** Write here your list of checking websites *(default "empty")*
  - `port` - **Array** Write here your list of websites port *(default "empty")*
  - `method` - **String** A string specifying the HTTP request method *(default "GET")*
  - `agent` - **String** User Agent for sending request *(default "monitode crawl")*
  - `file` - **String** Where save information of connection *(default "status")*
  - `timeout` - **Float** Timeout (second) for email send *(default "120")*

## Examples

Take a look at my [examples](https://github.com/hex7c0/monitode/tree/master/examples)

Or look at [pdf](https://github.com/hex7c0/monitode/raw/master/examples/monitode.pdf) file

### [License GPLv3](http://opensource.org/licenses/GPL-3.0)


History.md
v2.6.8 / 2015-01-22
==================

  * Package fix
  * _Deps_:
    * Update `express`@4.11.1
    * Update `express-sitemap`@1.4.0
    * Update `basic-authentication`@1.5.10
    * Update `server-signature`@1.0.9
    * Updata `mongodb`@2.0.14
  * _Web_:
    * Update `angular`@1.3.9

v2.6.7 / 2015-01-08
==================

  * _Deps_:
    * Update `logger-request`@3.2.7
    * Update `express`@4.10.7

v2.6.6 / 2014-12-31
==================

  * _Deps_:
    * Update `logger-request`@3.2.6
    * Use `mongodb`@2
  * _Web_:
    * Update `angular`@1.3.8

v2.6.5 / 2014-12-17
==================

  * _Deps_:
    * Update `express`@4.10.6

v2.6.4 / 2014-12-13
==================

  * _Deps_:
    * Update `timeout-request`@1.3.0
    * Update `logger-request`@3.2.3
    * Update `express`@4.10.5
    * Update `mongodb`@1.4.23
  * _Web_:
    * Update `angular`@1.3.6

v2.6.3 / 2014-11-23
==================

  * Update `express`@4.10.4
  * Update `mongodb`@1.4.22
  * Update `nano`@6.0.2
  * Update `angular`@1.3.4

v2.6.2 / 2014-11-23
==================

  * _Deps_:
    * Update `basic-authentication`@1.5.9
    * Update `express-sitemap`@1.3.8
    * Update `server-signature`@1.0.7
    * Update `logger-request`@3.2.2
    * Update `tickle`@1.1.8
  * _Web_:
    * Update `angular`@1.3.3

v2.6.1 / 2014-11-12
==================

  * Update `logger-request`@3.2.0
  * Update `timeout-request`@1.2.1
  * Update `express`@4.10.2
  * Update `nano`@5.12.2
  * Update `angular`@1.3.2

v2.6.0 / 2014-11-05
==================

  * Web Fix
  * _Deps_:
    * Update `logger-request`@3.1.1
    * Update `server-signature`@1.0.5
    * Update `timeout-request`@1.2.0
    * Update `express`@4.10.1
  * _Web_:
    * Update `angular`@1.3.1

v2.5.21 / 2014-10-29
==================

  * `jshint`
  * Update `logger-request`@3.0.13
  * Update `basic-authentication`@1.5.8
  * Update `tickle`@1.1.7
  * Update `express-sitemap`@1.3.7
  * Update `startline`@1.3.11
  * Update `timeout-request`@1.1.7
  * Using `server-signature`@1.0.4

v2.5.20 / 2014-10-26
==================

  * Update `logger-request`@3.0.11
  * Update `express`@4.10.0

v2.5.19 / 2014-10-22
==================

  * Update `timeout-request`@1.1.6

v2.5.18 / 2014-10-18
==================

  * Update `express`@4.9.8
  * Update `d3`@3.4.13

v2.5.17 / 2014-10-15
==================

  * Update `nano`@5.12.0
  * Update `angular`@1.3.0

v2.5.16 / 2014-10-11
==================

  * Update `logger-request`@3.0.10
  * Update `express`@4.9.7
  * Update `mongodb`@1.4.19
  * Update `angular`@1.3.0-rc.5
  * Update `d3`@3.4.12

v2.5.15 / 2014-10-04
==================

  * Remove jsdoc
  * `package.json` min

v2.5.14 / 2014-10-03
==================

  * Update `logger-request`@3.0.8
  * Update `mongodb`@1.4.18
  * Update `angular`@1.3.0-rc.4

v2.5.13 / 2014-09-25
==================

  * Update `mongodb`@1.4.17

v2.5.12 / 2014-09-25
==================

  * `GLOBAL` to `global`
  * Update `angular`@1.3.0-rc.3
  * Update `express`@4.9.5
  * Update `mongodb`@1.4.15

v2.5.11 / 2014-09-20
==================

  * Update `express`@4.9.4

v2.5.10 / 2014-09-19
==================

  * Update `express`@4.9.3

v2.5.9 / 2014-09-18
==================

  * Update `logger-request`@3.0.7
  * Update `express`@4.9.2
  * Update `mongodb`@1.4.12
  * Update `nano`@5.11.2

v2.5.8 / 2014-09-17
==================

  * Web modules error handling
  * Update `mongodb`@1.4.11
  * Update `express`@4.9.1
  * Update `angular`@1.3.0-rc.2

v2.5.7 / 2014-09-12
==================

  * Update `nodemailer`@1.3.0

v2.5.6 / 2014-09-10
==================

  * Update `express`@4.9.0
  * Update `startline`@1.3.9
  * Update `angular`@1.3.0-rc.1

v2.5.5 / 2014-09-08
==================

  * Update `nodemailer`@1.2.2
  * Update `startline`@1.3.8

v2.5.3 / 2014-09-05
==================

  * Update `express`@4.8.8
  * Update `mongodb`@1.4.10

v2.5.1 / 2014-09-01
==================

  * Avoid try-catch
  * Update `basic-authentication`@1.5.5
  * Update `logger-request`@3.0.4

v2.5.0 / 2014-08-30
==================

  * Bower package
  * Update `express`@4.8.7
  * Update `angular`@1.3.0-rc.0
  * Web Structure in /public

v2.4.16 / 2014-08-28
==================

  * Update `express`@4.8.6
  * Update `mongodb`@1.4.9
  * Update `nano`@5.11.0

v2.4.15 / 2014-08-23
==================

  * Performance tips
  * _Deps_:
    * Update `basic-authentication`@1.5.4
    * Update `express`@4.8.5
    * Update `express-sitemap`@1.3.4
    * Update `logger-request`@3.0.3
    * Update `nodemailer`@1.2.1
    * Update `startline`@1.3.5
    * Update `tickle`@1.1.4
    * Update `timeout-request`@1.1.3
  * _Web_:
    * Update `angular`@1.3.0-beta.19
    * Update `c3`@0.3.0

v2.4.14 / 2014-08-17
==================

  * Update `logger-request`@3.0.2

v2.4.12 / 2014-08-15
==================

  * Update `logger-request`@3.0.0
  * Update `express`@4.8.4

v2.4.11 / 2014-08-14
==================

  * Update `angularjs`@1.3.0-beta.18
  * Update `c3`@2.4.5

v2.4.10 / 2014-08-14
==================

  * Update `express-session`@1.3.3

v2.4.9 / 2014-08-11
==================

  * Update README.md
  * https://github.com/strongloop/express/issues/2290

v2.4.8 / 2014-08-08
==================

  * Update `express`@4.8.2

v2.4.7 / 2014-08-07
==================

  * Update `express`@4.8.1
  * Update `startline`@1.3.3

v2.4.6 / 2014-08-04
==================

  * Add "couch" (options)
  * Resolve path
  * Update `mongodb`@1.4.8
  * Update `basic-authentication`@1.5.1
  * Update `express-sitemap`@1.3.1
  * Update `logger-request`@2.2.3
  * Update `startline`@1.3.2
  * Update `tickle`@1.1.2
  * Update `timeout-request`@1.1.1

v2.4.5 / 2014-08-03
==================

  * Update `mongodb`@1.4.8
  * Don't use "wrap" for web script

v2.4.4 / 2014-07-31
==================

  * Fix clean GLOBAL var
  * Update `logger-request`@2.2.2
  * Update `nodemailer`@1.1.1

v2.4.3 / 2014-07-31
==================

  * Uglify angular script (mangle:false) + wrapper

v2.4.2 / 2014-07-29
==================


index.js
'use strict';
/**
 * @file monitode main
 * @module monitode
 * @package monitode
 * @subpackage main
 * @version 2.6.0
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

/*
 * initialize module
 */
// import
try {
    var resolve = require('path').resolve;
} catch (MODULE_NOT_FOUND) {
    console.error(MODULE_NOT_FOUND);
    process.exit(1);
}

/*
 * functions
 */
/**
 * option setting
 * 
 * @exports monitode
 * @function monitode
 * @param {Object} opt - various options. Check README.md
 * @return {Function}
 */
function monitode(opt) {

    var min = __dirname + '/min/';
    var spinterogeno = [];
    var options = opt || Object.create(null);
    // global
    /**
     * @global
     */
    var my = global.monitode = {
        min: min,
        output: Boolean(options.output),
        tickle: Boolean(options.tickle),
        app: false,
        os: Boolean(options.os),
        monitor: {
            os: true,
            log: true
        }
    };
    if (options.app) {
        my.app = {
            _router: {
                stack: options.app._router.stack,
            }
        };
    }
    if (my.os) {
        /**
         * @global
         */
        global.monitode.net = {
            inn: {
                pacs: 0,
                errs: 0,
            },
            out: {
                pacs: 0,
                errs: 0,
            },
        };
        /**
         * @global
         */
        global.monitode.io = {
            tps: 0,
            mbs: 0,
        };
    }

    // http
    options.http = options.http || Object.create(null);
    if (options.http.enabled === false ? false : true) {
        my.http = {
            port: Number(options.http.port) || 30000,
            user: String(options.http.user || 'admin'),
            password: String(options.http.password || 'password'),
            agent: String(options.http.agent || ''),
            realm: String(options.http.realm || 'Monitode'),
            file: options.http.file,
            hash: options.http.hash,
            dir: String(options.http.dir || __dirname + '/public/'),
        };
        spinterogeno.push(require(min + 'module/web.js'));
    }

    // https
    options.https = options.https || Object.create(null);
    if (options.https.key && options.https.cert) {
        my.https = {
            key: resolve(String(options.https.key)),
            cert: resolve(String(options.https.cert)),
            port: Number(options.https.port) || 30003,
            user: String(options.https.user || 'admin'),
            password: String(options.https.password || 'password'),
            agent: String(options.https.agent || ''),
            realm: String(options.https.realm || 'Monitode'),
            file: options.https.file,
            hash: options.https.hash,
            dir: String(options.https.dir || __dirname + '/public/'),
        };
        spinterogeno.push(require(min + 'module/webs.js'));
    }

    // logger
    options.logger = options.logger || Object.create(null);
    my.logger = Object.create(null);
    if (options.logger.file) {
        my.logger = {
            file: resolve(String(options.logger.file)),
            timeout: (parseFloat(options.logger.timeout) || 5) * 1000,
        };
        spinterogeno.push(require(min + 'module/file.js'));
    }
    if (options.logger.log) {
        my.logger.log = resolve(String(options.logger.log));
        /**
         * @global
         */
        global.monitode.log = {
            counter: 0,
            size: 0,
        };
        /**
         * @global
         */
        global.monitode.event = Object.create(null);
    }

    // database
    options.db = options.db || Object.create(null);
    if (options.db.mongo || options.db.couch) {
        my.db = {
            database: String(options.db.database || 'monitode'),
            timeout: (parseFloat(options.db.timeout) || 20) * 1000,
        };
        if ((my.db.mongo = options.db.mongo)) {
            spinterogeno.push(require(min + 'module/mongodb.js'));
        }
        if ((my.db.couch = options.db.couch)) {
            spinterogeno.push(require(min + 'module/couchdb.js'));
        }
    }

    // mail
    options.mail = options.mail || Object.create(null);
    if (options.mail.provider) {
        my.mail = {
            provider: String(options.mail.provider),
            user: String(options.mail.user || 'admin'),
            password: String(options.mail.password || 'password'),
            to: Array.isArray(options.mail.to) === true ? options.mail.to : [],
            subject: String(options.mail.subject || 'monitode report'),
            timeout: (parseFloat(options.mail.timeout) || 60) * 1000,
        };
        spinterogeno.push(require(min + 'module/mail.js'));
    }

    // status
    options.status = options.status || Object.create(null);
    if (Boolean(options.status.enabled)) {
        my.status = {
            site: Array.isArray(options.status.site) === true ? options.status.site
                    : [],
            port: Array.isArray(options.status.port) === true ? options.status.port
                    : [],
            method: String(options.status.method || 'GET'),
            agent: String(options.status.agent || 'monitode crawl'),
            file: resolve(String(options.status.file || 'status')),
            timeout: (parseFloat(options.status.timeout) || 120) * 1000,
        };
        spinterogeno.push(require(min + 'module/status.js'));
    }

    // start
    for (var i = 0, il = spinterogeno.length; i < il; i++) {
        spinterogeno[i]();
    }

    return function(req, res, next) {

        return next ? next() : null;
    };
}
module.exports = monitode;
